pkg/executor/infoschema_reader.go

```diff
@@ -563,151 +587,275 @@ func (e *memtableRetriever) updateStatsCacheIfNeed() bool {
 	}
-	extractor, ok := e.extractor.(*plannercore.InfoSchemaTablesExtractor)
-	if ok && extractor.SkipRequest {
-		return nil
+	schemas := is.AllSchemaNames()
+	slices.SortFunc(schemas, func(a, b model.CIStr) int {
+		return strings.Compare(a.L, b.L)
+	})
+	return schemas
+}
+
+func getMatchTableInfosForPartitions(
+	ctx context.Context,
+	extractor base.MemTablePredicateExtractor,
+	schema model.CIStr,
+	is infoschema.InfoSchema,
+) ([]*model.TableInfo, error) {
+	ex := extractor.(plannercore.TableSchemaSelector)
+	if ex == nil || !ex.HasTables() {
+		// There is no specified table in predicate.
+		return is.SchemaTableInfos(ctx, schema)
+	}
+	tables := make([]*model.TableInfo, 0, 8)
+	// Find all table infos from predicate.
+	for _, n := range ex.SelectedTableNames() {
+		tbl, err := is.TableByName(ctx, schema, n)
+		if err != nil {
+			if terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
+				continue
+			}
+			return nil, errors.Trace(err)
+		}
+		tables = append(tables, tbl.Meta())
 	}
-	for _, schema := range schemas {
-		if ok && extractor.Filter("table_schema", schema.L) {
+	for _, pid := range ex.SelectedPartitionIDs() {
+		tbl, db, _ := is.FindTableByPartitionID(pid)
+		if tbl == nil {
 			continue
 		}
-		tables, err := e.is.SchemaTableInfos(ctx, schema)
-		if err != nil {
-			return errors.Trace(err)
+		if db.Name.L != schema.L {
+			continue
 		}
-		for _, table := range tables {
-			if ok && extractor.Filter("table_name", table.Name.L) {
+		tables = append(tables, tbl.Meta())
+	}
+	return deduplicateTableInfos(tables), nil
+}
+
+func getMatchTableInfos(
+	ctx context.Context,
+	extractor base.MemTablePredicateExtractor,
+	schema model.CIStr,
+	is infoschema.InfoSchema,
+) ([]*model.TableInfo, error) {
+	ex := extractor.(plannercore.TableSchemaSelector)
+	if ex == nil || !ex.HasTables() {
+		// There is no specified table in predicate.
+		return is.SchemaTableInfos(ctx, schema)
+	}
+	tables := make([]*model.TableInfo, 0, 8)
+	// Find all table infos from predicate.
+	for _, n := range ex.SelectedTableNames() {
+		tbl, err := is.TableByName(ctx, schema, n)
+		if err != nil {
+			if terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
 				continue
 			}
-			collation := table.Collate
-			if collation == "" {
-				collation = mysql.DefaultCollationName
+			return nil, errors.Trace(err)
+		}
+		tables = append(tables, tbl.Meta())
+	}
+	for _, id := range ex.SelectedTableIDs() {
+		tbl, ok := is.TableByID(id)
+		if !ok {
+			continue
+		}
+		_, err := is.TableByName(ctx, schema, tbl.Meta().Name)
+		if err != nil {
+			if terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
+				continue
 			}
-			createTime := types.NewTime(types.FromGoTime(table.GetUpdateTime().In(loc)), createTimeTp, types.DefaultFsp)
+			return nil, errors.Trace(err)
+		}
+		tables = append(tables, tbl.Meta())
+	}
+	return deduplicateTableInfos(tables), nil
+}
 
-			createOptions := ""
+func deduplicateTableInfos(tables []*model.TableInfo) []*model.TableInfo {
```

Comment:
maybe at caller directly use `map[int64]*model.TableInfo` and use `maps.Values()` to get the slice to return.